tools/gengen/util/gengen.go

```diff
@@ -114,203 +118,74 @@ type RenderedMinerInfo struct {
-		StateRoot:       e.NewCid(stateRoot),
-		Messages:        e.NewCid(metaCid),
-		MessageReceipts: e.NewCid(emptyAMTCid),
-		BLSAggregateSig: crypto.Signature{
-			Type: crypto.SigTypeBLS,
-			Data: emptyBLSSignature[:],
-		},
-		Ticket:    consensus.GenesisTicket,
-		Timestamp: cfg.Time,
-	}
-
-	c, err := cst.Put(ctx, geneblk)
+	genCid, err := generator.genBlock(ctx)
 	if err != nil {
 		return nil, err
 	}
 
 	return &RenderedGenInfo{
-		Keys:       keys,
-		GenesisCid: c,
-		Miners:     miners,
+		Keys:       generator.keys,
+		GenesisCid: genCid,
+		Miners:     minerInfos,
 	}, nil
 }
 
-func genKeys(cfgkeys int, pnrg io.Reader) ([]*crypto.KeyInfo, error) {
-	keys := make([]*crypto.KeyInfo, cfgkeys)
-	for i := 0; i < cfgkeys; i++ {
-		ki := crypto.NewBLSKeyRandom() // TODO: use seed, https://github.com/filecoin-project/go-filecoin/issues/3781
-		keys[i] = &ki
-	}
-
-	return keys, nil
-}
-
-func setupPrealloc(vm consensus.GenesisVM, st *state.State, rnd crypto.RandomnessSource, keys []*crypto.KeyInfo, prealloc []string) error {
+// GenGenesisCar generates a car for the given genesis configuration
+func GenGenesisCar(cfg *GenesisCfg, out io.Writer) (*RenderedGenInfo, error) {
+	ctx := context.Background()
 
-	if len(keys) < len(prealloc) {
-		return fmt.Errorf("keys do not match prealloc")
-	}
+	bstore := blockstore.NewBlockstore(ds.NewMapDatastore())
+	bstore = blockstore.NewIdStore(bstore)
+	dserv := dag.NewDAGService(bserv.New(bstore, offline.Exchange(bstore)))
 
-	netact := actor.NewActor(builtin.AccountActorCodeID, abi.NewTokenAmount(1000000000000))
-	err := st.SetActor(context.Background(), builtin.RewardActorAddr, netact)
+	info, err := GenGen(ctx, cfg, bstore)
 	if err != nil {
-		return err
+		return nil, err
 	}
-
-	for i, v := range prealloc {
-		ki := keys[i]
-
-		addr, err := ki.Address()
-		if err != nil {
-			return err
-		}
-
-		valint, err := strconv.ParseUint(v, 10, 64)
-		if err != nil {
-			return err
-		}
-
-		_, err = vm.ApplyGenesisMessage(builtin.RewardActorAddr, addr,
-			builtin.MethodSend, abi.NewTokenAmount(int64(valint)), nil, rnd)
-		if err != nil {
-			return err
+	// Ignore cids that make it on chain but that should not be read through
+	// and therefore don't have corresponding blocks in store
+	ignore := cid.NewSet()
+	for _, m := range cfg.Miners {
+		for _, comm := range m.CommittedSectors {
+			ignore.Add(comm.CommR)
+			ignore.Add(comm.CommD)
+			ignore.Add(comm.DealCfg.CommP)
 		}
 	}
-	return nil
-}
-
-func setupMiners(vm consensus.GenesisVM, rnd crypto.RandomnessSource, keys []*crypto.KeyInfo, miners []*CreateStorageMinerConfig) ([]RenderedMinerInfo, error) {
-	var minfos []RenderedMinerInfo
 
-	for _, m := range miners {
-		addr, err := keys[m.Owner].Address()
-		if err != nil {
-			return nil, err
-		}
-
-		var pid peer.ID
-		if m.PeerID != "" {
-			p, err := peer.Decode(m.PeerID)
-			if err != nil {
-				return nil, err
+	ignoreWalkFunc := func(nd format.Node) (out []*format.Link, err error) {
```

Comment:
This is why we need an update to go-car, the current dependency can't do this for us.

Reply:
That's fine AFAIC. @frrist solved a similar problem in chain export - check for a re-usable piece of code there.